Content.Shared/Charges/Systems/SharedChargesSystem.cs

```diff
@@ -1,103 +1,216 @@
+        Dirty(action);
+    }
+
+    public bool TryUseCharge(Entity<LimitedChargesComponent?> entity)
+    {
+        return TryUseCharges(entity, 1);
+    }
+
+    public bool TryUseCharges(Entity<LimitedChargesComponent?> entity, int amount)
+    {
+        var current = GetCurrentCharges(entity);
+
+        if (current < amount)
+        {
             return false;
+        }
 
-        UseCharge(ent, ent.Comp);
+        AddCharges(entity, -amount);
         return true;
     }
 
+    [Pure]
+    public bool IsEmpty(Entity<LimitedChargesComponent?> entity)
+    {
+        return GetCurrentCharges(entity) == 0;
+    }
+
     /// <summary>
-    /// Gets the limited charges component and returns true if the number of charges remaining is less than the specified value.
-    /// Will return false if there is no limited charges component.
+    /// Resets action charges to MaxCharges.
     /// </summary>
-    public bool HasInsufficientCharges(EntityUid uid, int requiredCharges, LimitedChargesComponent? comp = null)
+    public void ResetCharges(Entity<LimitedChargesComponent?> action)
     {
-        // can't be empty if there are no limited charges
-        if (!Resolve(uid, ref comp, false))
-            return false;
+        if (!Resolve(action.Owner, ref action.Comp, false))
+            return;
 
-        return comp.Charges < requiredCharges;
+        var charges = GetCurrentCharges((action.Owner, action.Comp, null));
+
+        if (charges == action.Comp.MaxCharges)
+            return;
+
+        action.Comp.LastCharges = action.Comp.MaxCharges;
+        action.Comp.LastUpdate = _timing.CurTime;
+        Dirty(action);
+    }
+
+    public void SetCharges(Entity<LimitedChargesComponent?> action, int value)
+    {
+        action.Comp ??= EnsureComp<LimitedChargesComponent>(action.Owner);
+
+        var adjusted = Math.Clamp(value, 0, action.Comp.MaxCharges);
+
+        if (action.Comp.LastCharges == adjusted)
+        {
+            return;
+        }
+
+        action.Comp.LastCharges = adjusted;
+        action.Comp.LastUpdate = _timing.CurTime;
+        Dirty(action);
+    }
+
+    [Pure]
+    private TimeSpan GetNextRechargeTime(Entity<LimitedChargesComponent?, AutoRechargeComponent?> entity)
+    {
+        if (!Resolve(entity.Owner, ref entity.Comp1, ref entity.Comp2, false))
+        {
+            return TimeSpan.Zero;
+        }
+
+        // Okay so essentially we need to get recharge time to full, then modulus that by the recharge timer which should be the next tick.
+        var fullTime = ((entity.Comp1.MaxCharges - entity.Comp1.LastCharges) * entity.Comp2.RechargeDuration) + entity.Comp1.LastUpdate;
+        var timeRemaining = fullTime - _timing.CurTime;
+
+        if (timeRemaining < TimeSpan.Zero)
+        {
+            return TimeSpan.Zero;
+        }
+
+        var nextChargeTime = timeRemaining.TotalSeconds % entity.Comp2.RechargeDuration.TotalSeconds;
+        return TimeSpan.FromSeconds(nextChargeTime);
     }
```

Comment:
What happens right now is if this is called, it gets the charge time right?
But if another charge is used, this is reset.

So if you had limited charge to 3/3 and auto charge to 20 seconds and you used an action, the action goes to 2/3 and the timer starts. Works as expected.
If you fire off another charge while the timer is active, the timer will reset. So the action goes to 1/3 and the timer goes back to 20 seconds. The first charge still doesn't continue counting down and instead you need to wait another full 20 for the charge to recharge.

I don't think this is breaking enough for me to want to hold up the PR, but please put a TODO: with the bug information in this method so someone can look at it in the future.